### DISCUSSION TITLE
How does after_success condition work?

### DISCUSSION THREAD

**Discussion by gecko984** [Miksus/rocketry#220] [Q&A] [answered]
Hi,
(if Miksus reads it) first of all thank you for your great package! 
I have a small side project helping a friend of mine test his trading ideas.
I'm using a Rocketry app to 
- download stocks prices every 30 minutes, which takes about 7 minutes given my API limitations
- and run a script checking some conditions that the guy came up with every hour

It is vital that the script is only run after fresh data has been downloaded. So I wrote it like this, but it seems to work incorrectly, running the script before the download is complete

```python
@app.task(
    every('30 minutes') &
    time_of_day.between('10:00', '16:15') &  
    time_of_week.between("Mon", "Fri") &
    (time_of_hour.between("01:00", "02:00") | time_of_hour.between("31:00", "32:00"))
)
def task_download_data():
    download_data()

@app.task(
    every('1 hour') &
    after_success(task_download_data) &
    time_of_day.between('10:00', '17:00') &
    time_of_week.between("Mon", "Fri") &
    time_of_hour.after('01:00')
)
def task_script():
    run_script()

run_app()
```

Could you please give me a hint as to how do I change the conditions to ensure that task_script runs only after task_download_data is completed?

In general, I found that I was not able to get a good understanding from the docs (which are otherwise great) of how different conditions work in combination. So any insight or a mental framework for reasoning about them would be greatly appreciated too, because I'm pretty sure I'll keep using Rocketry as long as I write Python code :)

Thanks

**Comment by fdrigui** [comment]
Hello @gecko984,

As far I understood, there are 3 possible status for a task, with a different condition for each one:
- **Success**: when a task finishes and any error is raised.
- **Fail**: when a task finishes and some error is raised.
-  **Finish**: when a task finishes, whatever are the results.

1. First of all, import the conditions:
```python
from rocketry.conds import after_success, after_fail, after_finish
```

2. Configure your first task:
```python
@app.task(
    every('30 minutes') &
    time_of_day.between('10:00', '16:15') &  
    time_of_week.between("Mon", "Fri") &
    (time_of_hour.between("01:00", "02:00") | time_of_hour.between("31:00", "32:00"))
)
def task_download_data():
    download_data()
```

3. Create a condition based task, selecting one of [after_success, after_fail, after_finish]
```python
@app.task(after_finish(task_download_data))
def task_script():
    run_script()
```

For more information, check the documentation link: [https://rocketry.readthedocs.io/en/stable/tutorial/intermediate.html#pipelining](url)

**Comment by gecko984** [comment, marked as answer]
I read the docs more thoroughly and upon some reflection came to a solution using task return values, which I'll share here in case other people encounter similar problems. In my model example, the time scale is scaled down with a factor of 60, so basically instead of running the first task every 30 minutes, I run it every 30 seconds (for quicker debug obviously)

```python
import datetime
import random
from time import sleep

from rocketry import Rocketry
from rocketry.args import Return
from rocketry.conds import after_success, time_of_minute


def log(msg):
    print(f'{datetime.datetime.now().isoformat()} | {msg}')


app = Rocketry(execution='thread')


# note that there is no every(...), it is not necessary!
download_condition = time_of_minute.at('00') | time_of_minute.at('30')

@app.task(download_condition)
def task_download_data():
    # Get current second of minute and conclude whether this is the XX:00 run or XX:30 run.
    # Maybe you can do it with rocketry's tools without the datetime module, but I didn't figure out how.
    current_second = datetime.datetime.now().second
    is_00_run = current_second < 30
    
    # Simulate work
    log('downloading data...')
    sleep_duration = random.uniform(6, 9)
    sleep(sleep_duration)
    log('Finished downloading data')
    
    # We'll run or skip the second task depending on this task's return value.
    return {'is_00_run': is_00_run}


# note that there are no clock-related conditions here
script_condition = after_success(task_download_data)

@app.task(script_condition)
def task_run_script(arg=Return(task_download_data)):
    if arg['is_00_run']:
        log('It was the XX:00 download, running script')
        
        # simulate work
        sleep_duration = random.uniform(1, 2)
        sleep(sleep_duration)
        log('Finished running script')
    else:
        log('It was the XX:30 download, skipping script')


app.run()
```

And the output is

```
2023-10-20T02:50:00.026702 | downloading data...
2023-10-20T02:50:08.774167 | Finished downloading data
2023-10-20T02:50:08.803797 | It was the XX:00 download, running script
2023-10-20T02:50:10.338674 | Finished running script
2023-10-20T02:50:30.010709 | downloading data...
2023-10-20T02:50:36.837356 | Finished downloading data
2023-10-20T02:50:36.870338 | It was the XX:30 download, skipping script
2023-10-20T02:51:00.003473 | downloading data...
2023-10-20T02:51:07.878076 | Finished downloading data
2023-10-20T02:51:07.975875 | It was the XX:00 download, running script
2023-10-20T02:51:09.595374 | Finished running script
2023-10-20T02:51:30.098662 | downloading data...
2023-10-20T02:51:36.244605 | Finished downloading data
2023-10-20T02:51:36.253768 | It was the XX:30 download, skipping script
2023-10-20T02:52:00.100983 | downloading data...
2023-10-20T02:52:07.277265 | Finished downloading data
2023-10-20T02:52:07.367991 | It was the XX:00 download, running script
2023-10-20T02:52:08.570396 | Finished running script
```

So everything is working at expected - the script runs right after the download at XX minutes 00 seconds is complete, and doesn't run after the XX:30 downloads.

As a side note, I got rid of the `every` conditions. For one thing, they are not necessary here. For the other, they actually were harmful and here's why. `every('30 seconds')` forces the task to wait at least 30 seconds since the last run. And the tasks never start at the sharp time, as there's always some latency. 
Suppose the latency is exactly 0.2 seconds every time. Then the first run is at 00:00:00.**02**. The second run will be 00:00:30.**04**. The third will be at 00:01:00.**06**, etc.
It doesn't matter too much at first, but it can and will accumulate over long period of time, and your schedule will eventually get off the rails entirely.

So if you need a clock-bound schedule, you need to use `minutely` instead of `every('1 minute')` etc. The difference is explained in the docs here: https://rocketry.readthedocs.io/en/stable/handbooks/conditions/api/periodical.html#fixed-periods
I didn't figure out how to do it with periods that are multiples of 1 minute / 1 hour etc, like 30 seconds or 5 minutes though...

**Comment by gecko984** [comment]
One more solution (a simpler one) is to create a custom condition, that checks whether the first task is currently running, so that the second tasks can wait until the first task finishes. Still somewhat a crutch though. The condition looks like this:

```python
from rocketry.args import Task
app = Rocketry(execution='thread')

@app.task(time_of_minute.at(0) | time_of_minute.at(30)))
def first_task():
    sleep(5)

@app.cond()
def first_task_is_running(task=Task(first_task)):
    return task.is_running


```
